src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlDataContract.cs

```diff
@@ -131,34 +131,57 @@ public override XmlDictionaryString TopLevelElementNamespace
             { _helper.TopLevelElementNamespace = value; }
         }
 
-#if !NET_NATIVE
+#if NET_NATIVE
+        private CreateXmlSerializableDelegate _createXmlSerializableDelegate;
+        public CreateXmlSerializableDelegate CreateXmlSerializableDelegate        
+#else
         internal CreateXmlSerializableDelegate CreateXmlSerializableDelegate
+#endif
         {
+#if !NET_NATIVE
             /// <SecurityNote>
             /// Critical - fetches the critical CreateXmlSerializableDelegate property
             /// Safe - CreateXmlSerializableDelegate only needs to be protected for write; initialized in getter if null
             /// </SecurityNote>
             [SecuritySafeCritical]
             get
             {
-                if (_helper.CreateXmlSerializableDelegate == null)
+                if (DataContractSerializer.Option == SerializationOption.CodeGenOnly || DataContractSerializer.Option == SerializationOption.ReflectionAsBackup)
                 {
-                    lock (this)
+                    if (_helper.CreateXmlSerializableDelegate == null)
                     {
-                        if (_helper.CreateXmlSerializableDelegate == null)
+                        lock (this)
                         {
-                            CreateXmlSerializableDelegate tempCreateXmlSerializable = GenerateCreateXmlSerializableDelegate();
-                            Interlocked.MemoryBarrier();
-                            _helper.CreateXmlSerializableDelegate = tempCreateXmlSerializable;
+                            if (_helper.CreateXmlSerializableDelegate == null)
+                            {
+                                CreateXmlSerializableDelegate tempCreateXmlSerializable = GenerateCreateXmlSerializableDelegate();
+                                Interlocked.MemoryBarrier();
+                                _helper.CreateXmlSerializableDelegate = tempCreateXmlSerializable;
+                            }
                         }
                     }
+                    return _helper.CreateXmlSerializableDelegate;
                 }
-                return _helper.CreateXmlSerializableDelegate;
+
+                return () => ReflectionCreateXmlSerializable(this.UnderlyingType);
+            }
+#else              
+            get
+            {
+                if (DataContractSerializer.Option == SerializationOption.CodeGenOnly 
```

Comment:
Nit: I'd prefer to see explicit parentheses around the || and && conditions to make the precedence obvious.  Not blocking (as long as the implicit precedence is correct :-)